src/DataSource.js

```diff
@@ -119,27 +92,92 @@ prototype.fire = function(input) {
 prototype.pipeline = function(pipeline) {
   if (!arguments.length) return this._pipeline;
 
-  var ds = this;
+  this._inputNode = DataSourceInput(this);
+  this._outputNode = DataSourceOutput(this);
+
+  var graph = this._graph,
+      mutates = 0,
+      collector = this._inputNode,
+      i, node, router, collects;
+
+  for (i=0; i<pipeline.length; ++i) {
+    node = pipeline[i];
+
+    if (!node._collector && node.batch()) {
+      if (router) {
+        node = new Collector(graph);
+        pipeline.splice(i, 0, node);
+        router = false;
+      } else {
+        node._collector = collector;
+      }
+    }
+
+    if ((collects = node.collector())) collector = node;
+    router = router || node.router() && !collects;
+    mutates = mutates || node.mutates();
+  }
+  if (router) pipeline.push(collector = new Collector(graph));
+
+  pipeline.unshift(this._inputNode);
+  pipeline.push(this._outputNode);
+  this._collector = collector;
+  this._mutates = !!mutates;
+  this._graph.connect(this._pipeline = pipeline);
+  return this;
+};
+
+prototype.synchronize = function() {
+  var data = this._data, i, n;
 
-  // Add a collector to materialize the output of pipeline operators.
-  if (pipeline.length) {
-    ds._collector = new Collector(this._graph);
-    pipeline.push(ds._collector);
-    ds._revises = pipeline.some(function(p) { return p.revises(); });
-    ds._mutates = pipeline.some(function(p) { return p.mutates(); });
+  if (data.length && data[0]._prev) {
+    for (i=0, n=data.length; i<n; ++i) {
+      Tuple.prev_update(data[i]);
+    }
+  }
+  if (this._collector) {
```

Comment:
Still thinking through the implications of the new `synchronize` method. But, some room for further optimization here if the tuples in `this._collector` are the same ones in `this._data`?